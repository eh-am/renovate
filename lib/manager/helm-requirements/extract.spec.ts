import { fs } from '../../../test/util';
import { extractPackageFile } from './extract';

jest.mock('../../util/fs');

describe('lib/manager/helm-requirements/extract', () => {
  describe('extractPackageFile()', () => {
    beforeEach(() => {
      jest.resetAllMocks();
      fs.readLocalFile = jest.fn();
    });
    it('skips invalid registry urls', () => {
      fs.readLocalFile.mockResolvedValueOnce(`
      apiVersion: v1
      appVersion: "1.0"
      description: A Helm chart for Kubernetes
      name: example
      version: 0.1.0
      `);
      const content = `
      dependencies:
        - name: redis
          version: 0.9.0
          repository: '@placeholder'
        - name: postgresql
          version: 0.8.1
          repository: nope
        - name: broken
          version: 0.8.1
      `;
      const fileName = 'requirements.yaml';
      const result = extractPackageFile(content, fileName, {
        aliases: {
          stable: 'https://kubernetes-charts.storage.googleapis.com/',
        },
      });
      expect(result).not.toBeNull();
      expect(result).toMatchSnapshot();
      expect(result.deps.every((dep) => dep.skipReason)).toEqual(true);
    });
    it('parses simple requirements.yaml correctly', () => {
      fs.readLocalFile.mockResolvedValueOnce(`
      apiVersion: v1
      appVersion: "1.0"
      description: A Helm chart for Kubernetes
      name: example
      version: 0.1.0
      `);
      const content = `
      dependencies:
        - name: redis
          version: 0.9.0
          repository: https://kubernetes-charts.storage.googleapis.com/
        - name: postgresql
          version: 0.8.1
          repository: https://kubernetes-charts.storage.googleapis.com/
      `;
      const fileName = 'requirements.yaml';
      const result = extractPackageFile(content, fileName, {
        aliases: {
          stable: 'https://kubernetes-charts.storage.googleapis.com/',
        },
      });
      expect(result).not.toBeNull();
      expect(result).toMatchSnapshot();
    });
    it('parses simple requirements.yaml but skips if necessary fields missing', () => {
      fs.readLocalFile.mockResolvedValueOnce(`
      apiVersion: v1
      appVersion: "1.0"
      description: A Helm chart for Kubernetes
      name: example
      `);
      const fileName = 'requirements.yaml';
      const result = extractPackageFile('', fileName, {
        aliases: {
          stable: 'https://kubernetes-charts.storage.googleapis.com/',
        },
      });
      expect(result).toBeNull();
    });
    it('resolves aliased registry urls', () => {
      fs.readLocalFile.mockResolvedValueOnce(`
      apiVersion: v1
      appVersion: "1.0"
      description: A Helm chart for Kubernetes
      name: example
      version: 0.1.0
      `);
      const content = `
      dependencies:
        - name: redis
          version: 0.9.0
          repository: '@placeholder'
      `;
      const fileName = 'requirements.yaml';
      const result = extractPackageFile(content, fileName, {
        aliases: {
          placeholder: 'https://my-registry.gcr.io/',
        },
      });
      expect(result).not.toBeNull();
      expect(result).toMatchSnapshot();
      expect(result.deps.every((dep) => dep.skipReason)).toEqual(false);
    });
    it('skips local dependencies', () => {
      fs.readLocalFile.mockResolvedValueOnce(`
      apiVersion: v1
      appVersion: "1.0"
      description: A Helm chart for Kubernetes
      name: example
      version: 0.1.0
      `);
      const content = `
      dependencies:
        - name: redis
          version: 0.9.0
          repository: https://kubernetes-charts.storage.googleapis.com/
        - name: postgresql
          version: 0.8.1
          repository: file:///some/local/path/
      `;
      const fileName = 'requirements.yaml';
      const result = extractPackageFile(content, fileName, {
        aliases: {
          stable: 'https://kubernetes-charts.storage.googleapis.com/',
        },
      });
      expect(result).not.toBeNull();
      expect(result).toMatchSnapshot();
    });
    it('returns null if no dependencies', () => {
      fs.readLocalFile.mockResolvedValueOnce(`
      apiVersion: v1
      appVersion: "1.0"
      description: A Helm chart for Kubernetes
      name: example
      version: 0.1.0
      `);
      const content = `
      hello: world
      `;
      const fileName = 'requirements.yaml';
      const result = extractPackageFile(content, fileName, {
        aliases: {
          stable: 'https://kubernetes-charts.storage.googleapis.com/',
        },
      });
      expect(result).toBeNull();
    });
    it('returns null if requirements.yaml is invalid', () => {
      fs.readLocalFile.mockResolvedValueOnce(`
      apiVersion: v1
      appVersion: "1.0"
      description: A Helm chart for Kubernetes
      name: example
      version: 0.1.0
      `);
      const content = `
      Invalid requirements.yaml content.
      dependencies:
      [
      `;
      const fileName = 'requirements.yaml';
      const result = extractPackageFile(content, fileName, {
        aliases: {
          stable: 'https://kubernetes-charts.storage.googleapis.com/',
        },
      });
      expect(result).toBeNull();
    });
  });
});
