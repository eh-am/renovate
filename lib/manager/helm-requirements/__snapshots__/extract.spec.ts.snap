// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/manager/helm-requirements/extract extractPackageFile() parses simple requirements.yaml correctly 1`] = `
Object {
  "datasource": "helm",
  "deps": Array [
    Object {
      "currentValue": "0.9.0",
      "depName": "redis",
      "registryUrls": Array [
        "https://kubernetes-charts.storage.googleapis.com/",
      ],
    },
    Object {
      "currentValue": "0.8.1",
      "depName": "postgresql",
      "registryUrls": Array [
        "https://kubernetes-charts.storage.googleapis.com/",
      ],
    },
  ],
}
`;

exports[`lib/manager/helm-requirements/extract extractPackageFile() resolves aliased registry urls 1`] = `
Object {
  "datasource": "helm",
  "deps": Array [
    Object {
      "currentValue": "0.9.0",
      "depName": "redis",
      "registryUrls": Array [
        "https://my-registry.gcr.io/",
      ],
    },
  ],
}
`;

exports[`lib/manager/helm-requirements/extract extractPackageFile() skips invalid registry urls 1`] = `
Object {
  "datasource": "helm",
  "deps": Array [
    Object {
      "currentValue": "0.9.0",
      "depName": "redis",
      "registryUrls": Array [
        "@placeholder",
      ],
      "skipReason": "placeholder-url",
    },
    Object {
      "currentValue": "0.8.1",
      "depName": "postgresql",
      "registryUrls": Array [
        "nope",
      ],
      "skipReason": "invalid-url",
    },
    Object {
      "currentValue": "0.8.1",
      "depName": "broken",
      "skipReason": "no-repository",
    },
  ],
}
`;

exports[`lib/manager/helm-requirements/extract extractPackageFile() skips local dependencies 1`] = `
Object {
  "datasource": "helm",
  "deps": Array [
    Object {
      "currentValue": "0.9.0",
      "depName": "redis",
      "registryUrls": Array [
        "https://kubernetes-charts.storage.googleapis.com/",
      ],
    },
    Object {
      "currentValue": "0.8.1",
      "depName": "postgresql",
      "registryUrls": Array [
        "file:///some/local/path/",
      ],
      "skipReason": "local-dependency",
    },
  ],
}
`;

exports[`lib/manager/helm-requirements/extract extractPackageFile() skips only invalid dependences 1`] = `
Object {
  "datasource": "helm",
  "deps": Array [
    Object {
      "currentValue": undefined,
      "depName": "postgresql",
      "skipReason": "invalid-version",
    },
    Object {
      "currentValue": "0.0.1",
      "depName": undefined,
      "skipReason": "invalid-name",
    },
    Object {
      "currentValue": "0.0.1",
      "depName": "redis",
      "skipReason": "no-repository",
    },
    Object {
      "currentValue": "0.0.1",
      "depName": "redis",
      "registryUrls": Array [
        "https://kubernetes-charts.storage.googleapis.com/",
      ],
    },
  ],
}
`;

exports[`lib/manager/helm-requirements/extract extractPackageFile() validates name is present 1`] = `
Object {
  "datasource": "helm",
  "deps": Array [
    Object {
      "currentValue": "0.9.0",
      "depName": undefined,
      "skipReason": "invalid-name",
    },
  ],
}
`;

exports[`lib/manager/helm-requirements/extract extractPackageFile() validates name, repository and version are present 1`] = `
Object {
  "datasource": "helm",
  "deps": Array [
    Object {
      "currentValue": undefined,
      "depName": undefined,
      "skipReason": "invalid-version",
    },
  ],
}
`;

exports[`lib/manager/helm-requirements/extract extractPackageFile() validates repository is present 1`] = `
Object {
  "datasource": "helm",
  "deps": Array [
    Object {
      "currentValue": "0.0.0",
      "depName": "redis",
      "registryUrls": Array [
        "https://kubernetes-charts.storage.googleapis.com/",
      ],
    },
  ],
}
`;

exports[`lib/manager/helm-requirements/extract extractPackageFile() validates version is present 1`] = `
Object {
  "datasource": "helm",
  "deps": Array [
    Object {
      "currentValue": undefined,
      "depName": "redis",
      "skipReason": "invalid-version",
    },
  ],
}
`;
